internal/app/grpcclient/impl.go

```diff
@@ -113,24 +115,222 @@ func (i *invoker) unary(
 }
 
 func (i *invoker) serverStream(
-	_ context.Context,
-	_ *v1alpha1.ClientCompatRequest,
-) (*v1alpha1.ClientResponseResult, error) {
-	return nil, errors.New("server streaming is not yet implemented")
+	ctx context.Context,
+	ccr *v1alpha1.ClientCompatRequest,
+) (result *v1alpha1.ClientResponseResult, retErr error) {
+	result = &v1alpha1.ClientResponseResult{
+		ConnectErrorRaw: nil, // TODO
+	}
+
+	msg := ccr.RequestMessages[0]
+	req := &v1alpha1.ServerStreamRequest{}
+	if err := msg.UnmarshalTo(req); err != nil {
+		return nil, err
+	}
+	// Add the specified request headers to the request
+	ctx = grpcutil.AppendToOutgoingContext(ctx, ccr.RequestHeaders)
+
+	stream, err := i.client.ServerStream(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	// Read headers from the stream
+	hdr, err := stream.Header()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if result != nil {
+			// Set headers and trailers from the stream
+			result.ResponseHeaders = grpcutil.ConvertMetadataToProtoHeader(hdr)
+			result.ResponseTrailers = grpcutil.ConvertMetadataToProtoHeader(stream.Trailer())
+		}
+	}()
+
+	for {
+		msg, err := stream.Recv()
+		if err != nil {
+			if !errors.Is(err, io.EOF) {
+				result.Error = grpcutil.ConvertGrpcToProtoError(err)
+			}
+			break
+		}
+		// If the call was successful, get the returned payloads
+		// and the headers and trailers
+		result.Payloads = append(result.Payloads, msg.Payload)
+	}
+
+	return result, nil
 }
 
 func (i *invoker) clientStream(
-	_ context.Context,
-	_ *v1alpha1.ClientCompatRequest,
-) (*v1alpha1.ClientResponseResult, error) {
-	return nil, errors.New("client streaming is not yet implemented")
+	ctx context.Context,
+	ccr *v1alpha1.ClientCompatRequest,
+) (result *v1alpha1.ClientResponseResult, retErr error) {
+	result = &v1alpha1.ClientResponseResult{
+		ConnectErrorRaw: nil, // TODO
+	}
+
+	// Add the specified request headers to the request
+	ctx = grpcutil.AppendToOutgoingContext(ctx, ccr.RequestHeaders)
+
+	stream, err := i.client.ClientStream(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, msg := range ccr.RequestMessages {
+		csr := &v1alpha1.ClientStreamRequest{}
+		if err := msg.UnmarshalTo(csr); err != nil {
+			return nil, err
+		}
+
+		// Sleep for any specified delay
+		time.Sleep(time.Duration(ccr.RequestDelayMs) * time.Millisecond)
+
+		if err := stream.Send(csr); err != nil && errors.Is(err, io.EOF) {
+			break
+		}
+	}
+	resp, err := stream.CloseAndRecv()
+	if err != nil {
+		// If an error was returned, convert it to a gRPC error
+		result.Error = grpcutil.ConvertGrpcToProtoError(err)
+	} else {
+		// If the call was successful, get the returned payloads
+		result.Payloads = append(result.Payloads, resp.Payload)
+	}
+
+	hdr, err := stream.Header()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
```

Comment:
Not sure if it's worth it here since it's the last statement in the function 🤷‍♂️

Reply:
Probably not. Better to just inline it.